pubsub/gcloud/aio/pubsub/subscriber.py

```diff
@@ -169,56 +169,76 @@ async def consumer(
             max_tasks: int,
             nack_queue: 'Optional[asyncio.Queue[str]]',
             metrics_client: MetricsAgent) -> None:
-        semaphore = asyncio.Semaphore(max_tasks)
-        while True:
-            await semaphore.acquire()
+        try:
+            semaphore = asyncio.Semaphore(max_tasks)
+            while True:
+                await semaphore.acquire()
 
-            message, pulled_at = await message_queue.get()
+                message, pulled_at = await message_queue.get()
 
-            ack_deadline = await ack_deadline_cache.get()
-            if (time.perf_counter() - pulled_at) >= ack_deadline:
-                metrics_client.increment('pubsub.consumer.failfast')
-                message_queue.task_done()
-                semaphore.release()
-                continue
+                ack_deadline = await ack_deadline_cache.get()
+                if (time.perf_counter() - pulled_at) >= ack_deadline:
+                    metrics_client.increment('pubsub.consumer.failfast')
+                    message_queue.task_done()
+                    semaphore.release()
+                    continue
 
-            metrics_client.histogram(
-                'pubsub.consumer.latency.receive',
-                # publish_time is in UTC Zulu
-                # https://cloud.google.com/pubsub/docs/reference/rest/v1/PubsubMessage
-                time.time() - message.publish_time.timestamp())
-
-            task = asyncio.ensure_future(_execute_callback(
-                message,
-                callback,
-                ack_queue,
-                nack_queue,
-                metrics_client,
-            ))
-            task.add_done_callback(lambda _f: semaphore.release())
-            message_queue.task_done()
+                metrics_client.histogram(
+                    'pubsub.consumer.latency.receive',
+                    # publish_time is in UTC Zulu
+                    # https://cloud.google.com/pubsub/docs/reference/rest/v1/PubsubMessage
+                    time.time() - message.publish_time.timestamp())
+
+                task = asyncio.ensure_future(_execute_callback(
+                    message,
+                    callback,
+                    ack_queue,
+                    nack_queue,
+                    metrics_client,
+                ))
+                task.add_done_callback(lambda _f: semaphore.release())
+                message_queue.task_done()
+        except asyncio.CancelledError:
+            log.info('Consumer worker cancelled. Gracefully terminating...')
+            for _ in range(max_tasks):
+                await semaphore.acquire()
```

Comment:
So this relies on the producers having stopped pulling messages and cancelling the consumers themselves, correct?

Reply:
yes, except that consumers are stopped by top level `subscribe` after producers reported their proper shutdown